heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/CacheCore.java

```diff
@@ -0,0 +1,543 @@
+    long bucketId = makeBucketId(idx1, idx2);
+    LOG.info(Long.toHexString(bucketId) + " = "
+        + Integer.toHexString(idx1) + " " + Integer.toHexString(idx2));
+    // fetch the bucket
+    Map.Entry<Long, Map<Long, LinkedList<MetricDatapoint>>> entry =
+        cacheMetric.floorEntry(metricDatum.getTimestamp());
+    if (entry != null) {
+      Map<Long, LinkedList<MetricDatapoint>> locator = entry.getValue();
+      if (!locator.containsKey(bucketId)) {
+        locator.put(bucketId, new LinkedList<MetricDatapoint>());
+      }
+      LinkedList<MetricDatapoint> bucket = locator.get(bucketId);
+      // store the metric
+      MetricDatapoint datum = new MetricDatapoint();
+      datum.timestamp = metricDatum.getTimestamp();
+      datum.value = metricDatum.getValue();
+      bucket.offerFirst(datum);
+      LOG.info("addMetrics " + datum);
+    } else {
+      LOG.warning("too old metric: " + metricDatum);
+    }
+  }
+
+  private void addException(TopologyMaster.TmasterExceptionLog exceptionLog) {
+    String componentName = exceptionLog.getComponentName();
+    String instanceId = exceptionLog.getInstanceId();
+    assureComponentInstance(componentName, instanceId);
+    // get exception idx
+    int idx = idxComponentInstance.get(componentName).get(instanceId);
+    // fetch the bucket
+    if (!cacheException.containsKey(idx)) {
+      cacheException.put(idx, new LinkedList<ExceptionDatapoint>());
+    }
+    LinkedList<ExceptionDatapoint> bucket = cacheException.get(idx);
+    // store the exception
+    ExceptionDatapoint e = new ExceptionDatapoint();
+    e.componentName = exceptionLog.getComponentName();
+    e.hostname = exceptionLog.getHostname();
+    e.instanceId = exceptionLog.getInstanceId();
+    e.stackTrace = exceptionLog.getStacktrace();
+    e.lastTime = exceptionLog.getLasttime();
+    e.firstTime = exceptionLog.getFirsttime();
+    e.count = exceptionLog.getCount();
+    e.logging = exceptionLog.getLogging();
+    bucket.offerFirst(e);
+    // purge
+    while (bucket.size() > maxExceptionCount) {
+      LOG.warning("too many exception: " + exceptionLog);
+      bucket.pollLast();
+    }
+  }
+
+  /**
+   * for internal process use
+   *
+   * @param request idxMetricName == null: query all metrics
+   * idxMetricName == []: query none metric
+   * idxMetricName == [a, b, c .. ]: query metric a, b and c, ..
+   * <p>
+   * idxComponentInstance == null: query all components
+   * idxComponentInstance == []: query none component
+   * idxComponentInstance == [c1->null, ..]: query all instances of c1, ..
+   * idxComponentInstance == [c1->[], ..]: query none instance of c1, ..
+   * idxComponentInstance == [c1>[a, b, c, ..], ..]: query instance a, b, c, .. of c1, ..
+   * <p>
+   * assert: startTime <= endTime
+   */
+  public MetricsCacheQueryUtils.MetricResponse getMetrics(
+      MetricsCacheQueryUtils.MetricRequest request, MetricsFilter metricNameType) {
+    LOG.info(request.toString());
```

Comment:
please avoid logging objects with zero contextual information to the reader of the log re what the object is and why it's important to be logged. A log message like this without a custom message included can fill the logs and be very difficult to hunt down.